### PR TITLE
Patch/more5xx retries

### PR DESCRIPTION
**Pull request by jaketf**
One more before we cut a release.
see #71...

### CODE REVIEW

File: cdap/http.go [hunk 1]
```diff
@@ -44,13 +44,21 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
-		log.Print("retrying on intermittent 500 error in 2 seconds")
-		time.Sleep(2 * time.Second)
-		b, err = doHTTPCall(client, req)
-	}
+    // poor man's 5x EBO.
+    s := 2
+    for i:=0; i < 5; i++ {
+        if errors.As(err, &e) && e.code >= 500 && e.code < 600 {
+            log.Printf("retrying on intermittent 5xx error in %v seconds", s)
+            time.Sleep(2 * time.Second)
```
Comment (umairidris) [inline cdap/http.go]: I think you meant s * time.Second?

My personal preference would be to just do a linear 5 or 10 second sleep 2 more times. I have hit that 504 occassionally to and the very next call usually passes so I don't think we need 5 retries.

File: cdap/http.go [hunk 2]
```diff
@@ -44,13 +44,21 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
+            s *= 2
+            b, err = doHTTPCall(client, req)
+            if err != nil {
```
Comment (umairidris) [inline cdap/http.go]: I think you meant s * time.Second?

My personal preference would be to just do a linear 5 or 10 second sleep 2 more times. I have hit that 504 occassionally to and the very next call usually passes so I don't think we need 5 retries.

**Comment by umairidris** [comment]
I think you meant s * time.Second?

My personal preference would be to just do a linear 5 or 10 second sleep 2 more times. I have hit that 504 occassionally to and the very next call usually passes so I don't think we need 5 retries.

File: cdap/http.go [hunk 3]
```diff
@@ -44,13 +44,22 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
-	// CDAP REST intermittently returns 500 internal errors, we will retry on 500s once.
+	// CDAP REST intermittently returns 500, 504, etc. internal errors, we will retry on 5xxs once.
 	var e *httpError
-	if errors.As(err, &e) && e.code == 500 {
-		log.Print("retrying on intermittent 500 error in 2 seconds")
-		time.Sleep(2 * time.Second)
-		b, err = doHTTPCall(client, req)
-	}
+    // poor man's 3x EBO.
+    s := 2
```
Comment (umairidris) [inline cdap/http.go]: nit: var s time.Duration = 2

File: cdap/http.go [hunk 4]
```diff
@@ -44,13 +44,22 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
+    for i:=0; i < 3; i++ {
+        if errors.As(err, &e) && e.code >= 500 && e.code < 600 {
+            log.Printf("retrying on intermittent 5xx error in %v seconds", s)
+            // Have to explicitly cast if int var: https://play.golang.org/d0ZFLSVoAw
+            time.Sleep(time.Duration(s) * time.Second)
+            s *= 2
```
Comment (umairidris) [inline cdap/http.go]: time.Sleep(time.Duration(2*i) * time.Second) ? no ned for s anymore

Comment (jaketf) [reply cdap/http.go]: This would be linear not exponential backoff.

Comment (jaketf) [reply cdap/http.go]: I will use math.Pow(i, 2)

File: cdap/http.go [hunk 5]
```diff
@@ -44,12 +45,19 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
-		time.Sleep(2 * time.Second)
-		b, err = doHTTPCall(client, req)
+	// poor man's 3x EBO.
+	for i := 0; i < 3; i++ {
+		if errors.As(err, &e) && e.code >= 500 && e.code < 600 {
+			log.Printf("retrying on intermittent 5xx error in %v seconds", s)
+			// Have to explicitly cast if int var: https://play.golang.org/d0ZFLSVoAw
+			time.Sleep(time.Duration(math.Pow(2, i)) * time.Second)
+			b, err = doHTTPCall(client, req)
+			if err != nil {
```
Comment (umairidris) [inline cdap/http.go]: shouldn't this be err == nil?

File: cdap/http.go [hunk 6]
```diff
@@ -44,12 +45,21 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
-	// CDAP REST intermittently returns 500 internal errors, we will retry on 500s once.
+	// CDAP REST intermittently returns 500, 504, etc. internal errors, we will retry on 5xxs.
 	var e *httpError
-	if errors.As(err, &e) && e.code == 500 {
-		log.Print("retrying on intermittent 500 error in 2 seconds")
-		time.Sleep(2 * time.Second)
-		b, err = doHTTPCall(client, req)
+	// poor man's 3x EBO.
+	for i := 0; i < 3; i++ {
+		if errors.As(err, &e) && e.code >= 500 && e.code < 600 {
```
Comment (umairidris) [inline cdap/http.go]: It is preferred to make the breakout case in the if and move the rest outside i.e.

if !(errors.As(....)....) {
  break
}

log.Println(...)
time.Sleep(...)

File: cdap/http.go [hunk 7]
```diff
@@ -44,12 +45,21 @@ func httpCall(client *http.Client, req *http.Request) ([]byte, error) {
...
+			log.Print("retrying on intermittent 5xx error")
```
Comment (umairidris) [inline cdap/http.go]: log.Println, also remove the comment (it isn't needed as the compiler will catch this type of error anyway).

**Pull request merged**